pkg/controllers/clusterresourceplacement/controller_integration_test.go

```diff
@@ -0,0 +1,512 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+package clusterresourceplacement
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	placementv1beta1 "go.goms.io/fleet/apis/placement/v1beta1"
+	"go.goms.io/fleet/pkg/utils"
+)
+
+const (
+	member1Name = "member-1"
+	member2Name = "member-2"
+
+	eventuallyTimeout    = time.Minute * 2
+	consistentlyDuration = time.Second * 10
+	interval             = time.Millisecond * 250
+)
+
+var (
+	member1Namespace = fmt.Sprintf(utils.NamespaceNameFormat, member1Name)
+	member2Namespace = fmt.Sprintf(utils.NamespaceNameFormat, member2Name)
+
+	commonCmpOptions = cmp.Options{
+		cmpopts.IgnoreFields(metav1.ObjectMeta{}, "ResourceVersion", "UID", "CreationTimestamp", "ManagedFields", "Generation"),
+		cmpopts.IgnoreFields(metav1.OwnerReference{}, "UID"),
+	}
+	cmpPolicySnapshotOptions = cmp.Options{
+		commonCmpOptions,
+		cmpopts.IgnoreFields(placementv1beta1.ClusterSchedulingPolicySnapshot{}, "TypeMeta"),
+	}
+	cmpResourceSnapshotOptions = cmp.Options{
+		commonCmpOptions,
+		cmpopts.IgnoreFields(placementv1beta1.ClusterResourceSnapshot{}, "TypeMeta"),
+	}
+	cmpCRPOptions = cmp.Options{
+		commonCmpOptions,
+		cmpopts.IgnoreFields(placementv1beta1.ClusterResourcePlacement{}, "TypeMeta"),
+		cmpopts.IgnoreFields(metav1.Condition{}, "Message", "LastTransitionTime", "ObservedGeneration"),
+		cmpopts.SortSlices(func(c1, c2 metav1.Condition) bool {
+			return c1.Type < c2.Type
+		}),
+		cmpopts.SortSlices(func(c1, c2 placementv1beta1.ResourcePlacementStatus) bool {
+			return c1.ClusterName < c2.ClusterName
+		}),
+	}
+	crpOwnerReference = metav1.OwnerReference{
+		APIVersion:         fleetAPIVersion,
+		Kind:               "ClusterResourcePlacement",
+		Name:               testName,
+		Controller:         pointer.Bool(true),
+		BlockOwnerDeletion: pointer.Bool(true),
+	}
+)
+
+func validatePolicySnapshot(crp *placementv1beta1.ClusterResourcePlacement, want *placementv1beta1.ClusterSchedulingPolicySnapshot) *placementv1beta1.ClusterSchedulingPolicySnapshot {
```

Comment:
Some of these validate functions also return objects, which is a little bit confusing on my first read. It's just my personal opinion, but a name like `retrieveAndValidateXXX` would make things a little bit clearer.